# Necessary Libraries

In [1]:
import numpy as np
from sympy import Matrix
import matplotlib.pyplot as plt
import seaborn as sns


# Task 1 (Create my Personal Matrix)


In [7]:
# d1(last digit of my student ID) = 9
# d2(second last digit of my student ID) = 4

A = np.array([[11, 5], [18, 6]])



# Task 2

In [17]:
#1. shape of A
size_A = A.shape

# 2. Determinant of A
det_A = np.linalg.det(A)

# 3. Inverse of A
if np.linalg.det(A) != 0:
    inv_A = np.linalg.inv(A)
else:
    inv_A = "Matrix is singular, cannot compute inverse"

# 4. Rank of A
rank_A = np.linalg.matrix_rank(A)

# 5. Eigenvalues
eigenvalues, eigenvectors = np.linalg.eig(A)

#display results
print("Shape of A:", size_A)
print("Determinant of A:", det_A)
print("Inverse of A:\n", inv_A)
print("Rank of A:", rank_A)
print("Eigenvalues of A:", eigenvalues)
print("Eigenvectors of A:\n", eigenvectors)

Shape of A: (2, 2)
Determinant of A: -24.000000000000004
Inverse of A:
 [[-0.25        0.20833333]
 [ 0.75       -0.45833333]]
Rank of A: 2
Eigenvalues of A: [18.31070844 -1.31070844]
Eigenvectors of A:
 [[ 0.5645252  -0.37629789]
 [ 0.82541583  0.92649873]]


# Task 3(Value change experiment)

In [16]:
A = np.array([[11, 5], [18, 6]])

#Create a new matrix B
B = np.copy(A)
B[0][1] = B[0][1] + 1

print("B:\n", B)

B:
 [[11  6]
 [18  6]]


#Task 4(Re-analyze Matrix B)

In [18]:
#1. shape of B
size_B = B.shape

# 2. Determinant of B
det_B = np.linalg.det(B)

# 3. Inverse of B
if np.linalg.det(B) != 0:
    inv_B = np.linalg.inv(B)
else:
    inv_B = "Matrix is singular, cannot compute inverse"

# 4. Rank of B
rank_B = np.linalg.matrix_rank(A)

# 5. Eigenvalues
eigenvalues, eigenvectors = np.linalg.eig(B)

#display results
print("Shape of B:", size_B)
print("Determinant of B:", det_B)
print("Inverse of B:\n", inv_B)
print("Rank of B:", rank_B)
print("Eigenvalues of B:", eigenvalues)
print("Eigenvectors of B:\n", eigenvectors)

Shape of B: (2, 2)
Determinant of B: -42.00000000000001
Inverse of B:
 [[-0.14285714  0.14285714]
 [ 0.42857143 -0.26190476]]
Rank of B: 2
Eigenvalues of B: [19.18877916 -2.18877916]
Eigenvectors of B:
 [[ 0.59103662 -0.41409465]
 [ 0.80664473  0.91023383]]


In [ ]:
# Create two vectors of 10 random values each
x = np.random.rand(10) #by default [0,1) which includes 0 but excludes 1

y = np.random.rand(10)

# Comparison & Explanation

#1. How did the determinant change and why?
Ans: If a matrix A is = [(a,b), (c,d)]
 then the determinant of the matrix will be det(A) = ad - bc

 That's why if any value will change then the determinant will also be changed.

#2. Did the rank change? Explain.
Ans: After reanalyzing the matrix B, the rank didnot changed. Cause rank means the maximum number of linearly independent rows or columns that providing unique informations. So, if we see the martrix
B = [(11, 6), (18, 6)]
here the 1st row (11,6) and 2nd row (18,6) is unique means the rows providing unique equation or information. So, the rank didnot change or remain 2.


#3. How did the eigenvalues respond to the value change?
Ans:
After changing the value of matrix B the eigenvalues changed :-
The largest eigenvalue increased from 18.31 to 19.19 and
the smallest eigenvalue decreased(more negative) from -1.31 to -2.19.


#4. Is B easier or harder to invert than A? Why?
Ans: I think A is harder to invert than B cause A determinant(-24) is closer to zero than B determinant(-42). A determinant(-24) is closer to zero means the matrix is closer to singular. And singular matrix can't perform finding inverse operation. So, I mean A determinant(-24) making its inverse hard.

